rfd/0147-repo-mirror.md

```diff
@@ -0,0 +1,73 @@
+---
+authors: Mike Jensen (mike.jensen@goteleport.com)
+state: draft
+---
+
+# RFD 144 - Automatic Repo Mirroring
+
+## What
+
+This RFD outlines a method for continuously and automatically syncing a GitHub repository. This can be used to create private copies of a repository.
+
+## Why
+
+We currently face a few frictions which could be addressed through a more robust repository sync mechanism:
+1. Our internal `teleport-private` mirror of `teleport`, used for security development, currently requires manual updates to sync.
+2. Certain tools, like Dependabot, don't support scanning non-default branches, necessitating manual efforts to manage our release branches.
+3. Due to the committing of CI and scanning configurations in the repository, we receive duplicate alerts when using a repository for testing or as an internal mirror/fork.
+
+While addressing these immediate needs, the aim of this RFD is to establish a generic pattern for repository mirroring. This will provide flexibility for future requirements when syncing any upstream repository to a fork or mirror that we maintain.
+
+## Details
+
+The core requirements of this solution are:
+* Automatic and regular updates to a repository to ensure is's state matches the upstream.
+* The ability to make additional changes to the mirror repository, separate from the upstream repository (e.g., disabling actions we don't want to run on the mirror, or enabling additional scanning/actions).
+
+### Branch Structure
+
+* `master` (configurable) - This will serve as the repository's default branch. It will contain both the upstream history and any commits necessary for the synchronization process or other custom changes made to the mirror repository.
+* `sync/upstream-[UPSTREAM_BRANCH_NAME]` - This branch will be an exact copy of the upstream branch. Should there be a need to create a PR in this repository, the history should be based on this branch, not on `master`.
```

Comment:
> Should there be a need to create a PR in this repository

I would expect PRs to land against sync/rebase, as it holds other custom changes.

Edit: it seems that this instruction is again specific to teleport-private, in particular for the "fix" PRs that never actually land on any branch?

Reply:
You're correct, it would make sense to open a PR to `sync/rebase` if you needed to make additional changes to the rebase process.

I expect that to be a rare condition compared to the need to develop private fixes (which is what I am describing here)

Let me know if you see a way to word this better (or perhaps I should just remove any discussion of PR's since that discussed below also).

Reply:
Our other threads cover this discussion well, so closing.